File: lib/face_native/ios/Classes/data/DataModels.swift
import ObjectBox

// objectbox: entity
class FaceImageRecord {
    // Primary key of FaceImageRecord
    var id: Id = 0
    
    // empId references the PersonRecord primary key
    // objectbox: index
    var empId: Int64 = 0
    
    var personName: String = ""
    
    // The FaceNet model provides a 128-dimensional embedding
    // objectbox:hnswIndex: dimensions=128, distanceType=cosine, indexingSearchCount=400
    var faceEmbedding: [Float] = []
    
    required init() {} // Required by ObjectBox
    
    convenience init(id: Id = 0, empId: Int64, personName: String, faceEmbedding: [Float]) {
        self.init()
        self.id = id
        self.empId = empId
        self.personName = personName
        self.faceEmbedding = faceEmbedding
    }
    
    func toMap() -> [String: Any] {
        return [
            "employeeId": empId,
            "personName": personName,
            "faceEmbedding": faceEmbedding
        ]
    }
}

// Recognition metrics structure (not an entity)
struct RecognitionMetrics {
    let timeFaceDetection: Int64
    let timeVectorSearch: Int64
    let timeFaceEmbedding: Int64
    let timeFaceSpoofDetection: Int64
}

File: lib/face_native/android/src/main/kotlin/com/example/face_native/data/DataModels.kt
package com.example.face_native.data

import io.objectbox.annotation.Entity
 import io.objectbox.annotation.Uid
import io.objectbox.annotation.HnswIndex
import io.objectbox.annotation.Id
import io.objectbox.annotation.Index
import io.objectbox.annotation.VectorDistanceType

@Entity
data class FaceImageRecord(
    // primary-key of `FaceImageRecord`
    @Id var recordID: Long = 0,
    // empId references the PersonRecord primary key
    @Index var empId: Long = 0,
    var personName: String = "",
    // the FaceNet-512 model provides a 512-dimensional embedding
    // the FaceNet model provides a 128-dimensional embedding
    @HnswIndex(
        dimensions = 128,
        distanceType = VectorDistanceType.COSINE,
        indexingSearchCount=400,
    ) var faceEmbedding: FloatArray = floatArrayOf(),
)


data class RecognitionMetrics(
    val timeFaceDetection: Long,
    val timeVectorSearch: Long,
    val timeFaceEmbedding: Long,
    val timeFaceSpoofDetection: Long,
)
